mithril-aggregator/src/services/usage_reporter.rs

```diff
@@ -0,0 +1,307 @@
+use std::{collections::HashMap, sync::Arc, time::Duration};
+
+use crate::{
+    event_store::{EventMessage, TransmitterService},
+    MetricsService,
+};
+use chrono::{DateTime, Utc};
+use mithril_common::logging::LoggerExtensions;
+use serde::{Deserialize, Serialize};
+use slog::{info, Logger};
+
+/// Message sent to the event store to report a metric value.
+#[derive(Serialize, Deserialize)]
+struct MetricMessage {
+    counter: i64,
+    duration: Duration,
+    date: DateTime<Utc>,
+}
+/// Reporter of metrics about the usage of the application.
+pub struct UsageReporter {
+    transmitter_service: Arc<TransmitterService<EventMessage>>,
+    metric_service: Arc<MetricsService>,
+    last_metrics: HashMap<String, u32>,
+    logger: Logger,
+}
+
+impl UsageReporter {
+    /// Create a new UsageReporter.
+    pub fn new(
+        transmitter_service: Arc<TransmitterService<EventMessage>>,
+        metric_service: Arc<MetricsService>,
+        logger: Logger,
+    ) -> Self {
+        Self {
+            transmitter_service,
+            metric_service,
+            last_metrics: HashMap::new(),
+            logger: logger.new_with_component_name::<Self>(),
+        }
+    }
+
+    fn metrics_delta(
+        metrics_before: &HashMap<String, u32>,
+        metrics_after: &HashMap<String, u32>,
+    ) -> HashMap<String, i64> {
+        metrics_after
+            .iter()
+            .map(|(name, value)| {
+                let last_value = metrics_before.get(name).unwrap_or(&0);
+                let delta: i64 = (*value as i64) - (*last_value as i64);
+                (name.clone(), delta)
+            })
+            .filter(|(_name, value)| *value != 0)
```

Comment:
The `unwrap_or` is for getting last metric that could not exist in the map. 
The `filter` is on the difference between new and old value.
We may replace the `unwrap_or` using `if let Some` or other syntax but I'm not sure it's better. 
```
                let last_value = if let Some(value) = metrics_before.get(name) {
                    value
                } else {
                    &0
                };
```

We can use the `filter_map` to filter metric values to 0 with the following code but I prefer to compute values in `map` and then apply `filter` on the result. in my mind, the two notions are on different levels. The filter is a business rule to avoid too many events while the delta computation is a technical stuff. 
```
    if delta != 0 {
        Some((name.clone(), delta))
    } else {
        None
    }
```